core/plugins/org.polarsys.capella.core.platform.sirius.ui.actions/src/org/polarsys/capella/core/platform/sirius/ui/commands/ExchangeItemAllocationOnPortsCommand.java

```diff
@@ -105,41 +104,37 @@ protected Collection<ModelElement> retrieveModelElements(ModelElement modelEleme
   }
 
   @Override
-  protected void process(ModelElement element) {
+  protected boolean process(ModelElement element) {
     Collection<ExchangeItem> flows;
 
     if (element instanceof FunctionInputPort) {
       FunctionInputPort fp = (FunctionInputPort) element;
-      processPort(fp, getLinkedItems(fp));
+      return processPort(fp, getLinkedItems(fp));
 
     } else if (element instanceof FunctionOutputPort) {
       FunctionOutputPort fp = (FunctionOutputPort) element;
-      processPort(fp, getLinkedItems(fp));
+      return processPort(fp, getLinkedItems(fp));
 
     } else if (element instanceof FunctionalExchange) {
       FunctionalExchange fe = (FunctionalExchange) element;
       flows = ((FunctionalExchange) element).getExchangedItems();
 
-      if (fe.getSource() != null) {
-        if (fe.getSource() instanceof FunctionOutputPort) {
-          processPort((FunctionOutputPort) fe.getSource(), flows);
-        }
+      if (fe.getSource() instanceof FunctionOutputPort) {
+        return processPort((FunctionOutputPort) fe.getSource(), flows);
       }
-
-      if (fe.getTarget() != null) {
-        if (fe.getTarget() instanceof FunctionInputPort) {
-          processPort((FunctionInputPort) fe.getTarget(), flows);
-        }
+      if (fe.getTarget() instanceof FunctionInputPort) {
+        return processPort((FunctionInputPort) fe.getTarget(), flows);
```

Comment:
for an exchange, maybe do both no ? 
now it processPort only on the source, no longer on the target

Reply:
i removed the fe.getSource() != null because it is useless, instanceof will take care of checking if it null.
i didnt change the logic
same fot fe.getTarget()

Reply:
the **return** on the line 123 for the source, makes that it is no longer called for the target.

Reply:
done